clicommand/agent_pause.go

```diff
@@ -0,0 +1,122 @@
+package clicommand
+
+import (
+	"context"
+	"fmt"
+
+	"time"
+
+	"github.com/buildkite/agent/v3/api"
+	"github.com/buildkite/agent/v3/logger"
+	"github.com/buildkite/roko"
+	"github.com/urfave/cli"
+)
+
+const pauseDescription = `Usage:
+
+    buildkite-agent pause [options...]
+
+Description:
+
+Pauses the current agent.
+
+Example:
+
+    # Pauses the agent
+    $ buildkite-agent pause
+
+    # Pauses the agent with an explanatory note and a custom timeout
+    $ buildkite-agent pause --note 'too many llamas' --timeout-in-minutes 60`
+
+type AgentPauseConfig struct {
+	Note             string `cli:"note"`
+	TimeoutInMinutes int    `cli:"timeout-in-minutes"`
+
+	// Global flags
+	Debug       bool     `cli:"debug"`
+	LogLevel    string   `cli:"log-level"`
+	NoColor     bool     `cli:"no-color"`
+	Experiments []string `cli:"experiment" normalize:"list"`
+	Profile     string   `cli:"profile"`
+
+	// API config
+	DebugHTTP        bool   `cli:"debug-http"`
+	AgentAccessToken string `cli:"agent-access-token" validate:"required"`
+	Endpoint         string `cli:"endpoint" validate:"required"`
+	NoHTTP2          bool   `cli:"no-http2"`
+}
+
+var AgentPauseCommand = cli.Command{
+	Name:        "pause",
+	Usage:       "Pause the agent",
+	Description: pauseDescription,
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:   "note",
+			Usage:  "A descriptive note to record why the agent is paused",
+			EnvVar: "BUILDKITE_AGENT_PAUSE_NOTE",
+		},
+		cli.IntFlag{
+			Name:   "timeout-in-minutes",
+			Usage:  "Timeout after which the agent is automatically resumed, in minutes",
+			EnvVar: "BUILDKITE_AGENT_PAUSE_TIMEOUT_MINUTES",
+		},
+
+		// API Flags
+		AgentAccessTokenFlag,
+		EndpointFlag,
+		NoHTTP2Flag,
+		DebugHTTPFlag,
+
+		// Global flags
+		NoColorFlag,
+		DebugFlag,
+		LogLevelFlag,
+		ExperimentsFlag,
+		ProfileFlag,
+	},
+	Action: func(c *cli.Context) error {
+		ctx := context.Background()
+		ctx, cfg, l, _, done := setupLoggerAndConfig[AgentPauseConfig](ctx, c)
+		defer done()
+
+		return pause(ctx, cfg, l)
+	},
+}
+
+func pause(ctx context.Context, cfg AgentPauseConfig, l logger.Logger) error {
+	// Create the API client
+	client := api.NewClient(l, loadAPIClientConfig(cfg, "AgentAccessToken"))
+
+	// Retry the build cancellation a few times before giving up
```

Comment:
Is this talk of cancelation here accurate, or copy paste from the agent stop command?

I'd need to check the backend code, but IIRC pausing doesn't cancel the job/build but rather prevents new jobs being dispatched to it? Or did I misunderstand the whole thing?

Reply:
Just copypasta PEBKAC error.

Reply:
All good! The three commands are all basically the same, but I think the simplicity of copypasta is better than premature abstraction / DRY-ing up any of this stuff, so I 💯 would have ended up copying the same comments 😅